websock/session.nim

```diff
@@ -116,13 +119,86 @@ proc send*(
     else:
       default(MaskKey)
 
-  if opcode in {Opcode.Text, Opcode.Cont, Opcode.Binary}:
-    await ws.writeMessage(
-      data, opcode, maskKey, ws.extensions)
+  if opcode.isControl:
+    await ws.writeControl(data, opcode, maskKey)
+  else:
+    await ws.writeMessage(data, opcode, maskKey, ws.extensions)
 
-    return
+proc doSend(
+  ws: WSSession,
+  data: seq[byte] = @[],
+  opcode: Opcode,
+  fut: Future[void]): Future[void]
+  {.raises: [Defect, WSClosedError].} =
+  if fut.cancelled:
+    let sendFut = newFuture[void]("doSend")
+    sendFut.cancel()
+    return sendFut
+
+  let sendFut = ws.nonCancellableSend(data, opcode)
+
+  proc handleSent(future: pointer) =
+    if fut.finished:
+      return
+    if sendFut.failed:
+      fut.fail(sendFut.error)
+    elif sendFut.cancelled:
+      fut.cancel()
+    else:
+      fut.complete()
+
+  sendFut.addCallback(handleSent)
+
+  sendFut
+
+proc continueSending(ws: WSSession) {.gcsafe.} =
+  while ws.sendQueue.len > 0:
+    let
+      task = ws.sendQueue.popFirst()
+      fut = task.fut
+      sendFut =
+        try:
+          ws.doSend(task.data, task.opcode, fut)
+        except WSClosedError as exc:
+          fut.fail(exc)
+          continue
+
+    ws.sendFut = sendFut
 
-  await ws.writeControl(data, opcode, maskKey)
+    proc handleSent(future: pointer) =
+      ws.sendFut = nil
+      ws.continueSending()
+
+    ws.sendFut.addCallback(handleSent)
+    break
+
+proc send*(
+  ws: WSSession,
+  data: seq[byte] = @[],
+  opcode: Opcode): Future[void]
+  {.raises: [Defect, WSClosedError].} =
+  let fut = newFuture[void]("send")
+
+  if opcode.isControl:
+    # Control frames (see Section 5.5) MAY be injected in the middle of
+    # a fragmented message.  Control frames themselves MUST NOT be
+    # fragmented.
+    # See RFC 6455 Section 5.4 Fragmentation
+    discard ws.doSend(data, opcode, fut)
+    return fut
+
+  if ws.sendFut != nil:
```

Comment:
Might be easier to do `if ws.sendFut != nil or ws.sendFut.finished` (not need to set it to nil afterwards)

Reply:
Setting it to `nil` releases associated memory though.

Reply:
And when `ws.sendFut.finished`, it may still be necessary to go through `addLast` to ensure that potential continuations in the queue are also scheduled.

Reply:
> Setting it to nil releases associated memory though.

When finished, a Future will already release the closure environment on it's own: https://github.com/status-im/nim-chronos/blob/3e4b47e60a6033ec6732479b308982648ebe6074/chronos/asyncfutures2.nim#L373

Leaving only a few bytes behind

Reply:
Ah, that's good to know. Thanks for the pointer!